Refactor CMSIS

stephanosio opened the pull request:
At the time of writing, the Zephyr repository only includes CMSIS-Core(M) variant under ext/hal/cmsis and assumes that the only form of "CMSIS" available is CMSIS-Core(M).

In order to facilitate the development of Cortex-A and Cortex-R ports, the CMSIS-to-Zephyr RTOS integration structure should be refactored to support multiple CMSIS-Core variants as well as other coprocessor and extension variants.

For more details, see the issue #19717.

stephanosio committed:
ext: hal: cmsis: Relocate CMSIS-Core(M) files to its own directory.

This commit relocates the CMSIS-Core(M) Include directory that
currently resides directly under ext/hal/cmsis directory to its own
directory, Core, in order to allow other CMSIS variants to be added.

The name of CMSIS-Core(M) directory, Core, is following the original
name used by the upstream CMSIS repository.

For more details, see issue #19717.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
ext: hal: cmsis: Add HAS_CMSIS_CORE configuration.

The existing implementation used HAS_CMSIS configuration to specify
that CMSIS-Core(M) is used; when, in fact, there are other CMSIS
variants available such as CMSIS-Core(A) and CMSIS-DSP available.

This commit replaces the existing HAS_CMSIS configuration with
HAS_CMSIS_CORE to clarify that CMSIS-Core is used. It also introduces
the CMSIS-Core variant configuration, HAS_CMSIS_CORE_M, that is
automatically selected when HAS_CMSIS_CORE is enabled.

For more details, see issue #19717.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
ext: hal: cmsis: Update references to HAS_CMSIS to HAS_CMSIS_CORE.

This commit updates all references to HAS_CMSIS to use HAS_CMSIS_CORE
instead. With the changes introduced to allow multiple CMSIS variants
to be specified, the latter is semantically equivalent to the former.

For more details, see issue #19717.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

ioannisg commented on ext/hal/cmsis/README:
```diff
@@ -1,15 +1,15 @@
 
-The current version supported in Zephyr is 
+The current version supported in Zephyr is
 https://github.com/ARM-software/CMSIS_5/releases/tag/5.5.1
```
@galak @stephanosio we may have missed that in the last CMSIS upgrade. We have version 5.6.0, AFAIK.
Could you add a separate commit to fix the version in this text?

stephanosio replied:
@ioannisg Just added a commit to fix that: c253778e649919b38e097128615e836b0a470e43

ioannisg commented on soc/arm/nordic_nrf/nrf51/Kconfig.series:
```diff
@@ -12,7 +12,7 @@ config  SOC_SERIES_NRF51X
 	select HAS_SYS_POWER_STATE_DEEP_SLEEP_1
 	select XIP
-	select HAS_CMSIS
+	select HAS_CMSIS_CORE
```
Could we have the same thing for Nordic, as for the other ARM vendors? I.e. move the selection of CMSIS in the modules' Kconfig?

stephanosio replied:
Technically, `select HAS_CMSIS_CORE` is not really needed anywhere other than in arch/arm/core/Kconfig because `CPU_CORTEX_M` already selects it.

I think these redundant selections could use a clean-up in the future (after all, since arch/arm/core relies on CMSIS-Core, so this makes sense).

ioannisg replied:
OH, I see, then could you clean them up in this PR, if possible?

stephanosio replied:
> OH, I see, then could you clean them up in this PR, if possible?

On second thought, one could probably argue that `HAS_CMSIS_CORE` should be selected by HAL modules because they actually use and rely on CMSIS-Core and, while it is true that such selections are redundant due to `CPU_CORTEX_M` selecting it already on Cortex-M HALs, purely from dependency perspective, `HAS_CMSIS_CORE` should be selected by them.

In response to that argument, I think, before doing this clean-up, we need to clarify that using `CPU_CORTEX_M` predicates the use of CMSIS-Core as this is an essential component which Cortex-M port cannot function without and therefore it is absolutely safe to assume that `CPU_CORTEX_M` will always select `HAS_CMSIS_CORE`.

If there's no counterargument to this, I add a commit for this clean-up tomorrow.

ioannisg replied:
All I want is consistency. Right now Nordic selects the CMSIS in the SOC definition, while other vendors select it in the module Kconfig. So, I think we can just make nordic look like the others.

I agree that HAS_CMSIS should still be part of the vendor module Kconfig; this actually signifies that the HAL headers _do_ use the CMSIS headers. In theory, this does not need to be the case.

stephanosio replied:
A major `HAS_CMSIS_CORE` usage clean-up involving redundant selection removal and/or other forms of refactoring can be done later in a separate PR (preferably after Cortex-R and Cortex-A port are added, as this would probably bring in quite a few big changes).

For now, I have added a `HAS_CMSIS_CORE` usage clean-up only for consistency.

ioannisg replied:
I like that the CMSIS_CORE is now selected in NRFX,instead of nRF SOC SERIES.

ioannisg reviewed (approved):
It looks ok to me.
I added a few comments for additional commits that are actually unrelated to the patch. So @stephanosio I am fine if you don't want to do the changes in this PR.

stephanosio committed:
ext: hal: cmsis: Fix version information in README.

This commit fixes the incorrect CMSIS version information in the
README file.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
soc: arm: Clean up HAS_CMSIS_CORE usage.

For nordic_nrf, this commit relocates HAS_CMSIS_CORE selection from
SoC Kconfig to the HAL module Kconfig, as done for other SoCs.

For nxp_kinetis, remove redundant HAS_CMSIS_CORE selection in SoC
Kconfig, as it is already selected by the HAL Kconfig.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

galak commented on ext/hal/cmsis/Kconfig:
```diff
@@ -1,8 +1,10 @@
-config HAS_CMSIS
+config HAS_CMSIS_CORE
 	bool
+	select HAS_CMSIS_CORE_M if CPU_CORTEX_M
```
Hmm, wonder if this is better as 'default if CPU_CORTEX_M' instead of a select.

@ulfalizer thoughts?

stephanosio replied:
To elaborate, the idea here is that if `HAS_CMSIS_CORE` should automatically select an adequate CMSIS-Core type based on the CPU type.

For example, when CMSIS-Core(R) is added in the next PR, it would look something like this:
```
config HAS_CMSIS_CORE
	bool
	select HAS_CMSIS_CORE_M if CPU_CORTEX_M
	select HAS_CMSIS_CORE_R if CPU_CORTEX_R
```

The rationale for using `select` instead of `default` is as follows:

1. Since CMSIS-Core is a "driver" for CPU core, it makes very little sense to select a type that does not match the CPU type. For instance, if you have `CPU_CORTEX_M` selected, you would always want to select `HAS_CMSIS_CORE_M` which enables CMSIS-Core(M). For this reason, other config files should never need to override this.

2. Once `HAS_CMSIS_CORE` is selected, the right type of CMSIS-Core must be enabled. For example, it makes no sense to select `HAS_CMSIS_CORE` and then do `HAS_CMSIS_CORE_M=n` in another config file.

ioannisg replied:
This design solves another problem. Assume a vendor with all Cortex-A, Cortex-M and Cortex-R support. The vendor module will select HAS_CMSIS in the root module Kconfig file, without having to care about the different ARM variants. The right type will be selected based on the actual CPU variant. 

So I like the solution by proposed here by @stephanosio .

Of course, these options must be used somewhere.

galak reviewed (commented):
Generally looks good, one question about how to structure the Kconfig.

ioannisg commented:
> Generally looks good, one question about how to structure the Kconfig.

IMHO, HAS_CMSIS_CORE should not be selected by CPU_CORTEX_M. The option should signify that the vendor headers include the CMSIS headers. 

In theory, there could be ARM vendor HAL headers that do not include CMSIS and define own custom ARM headers.

stephanosio commented:
> IMHO, HAS_CMSIS_CORE should not be selected by CPU_CORTEX_M. The option should signify that the vendor headers include the CMSIS headers.

The problem with not selecting `HAS_CMSIS_CORE` for `CPU_CORTEX_M` is that arch/arm/core directly makes use of CMSIS-Core in places like the following:

https://github.com/zephyrproject-rtos/zephyr/blob/f79fbac101f8febb459acee577e9f6b76e00acc0/arch/arm/core/prep_c.c#L91-L97

Since `CPU_CORTEX_M` signifies that ARM Cortex-M arch is enabled and this arch implementation depends on CMSIS-Core, `HAS_CMSIS_CORE` should be selected by it.

ioannisg commented:
> > IMHO, HAS_CMSIS_CORE should not be selected by CPU_CORTEX_M. The option should signify that the vendor headers include the CMSIS headers.
> 
> The problem with not selecting `HAS_CMSIS_CORE` for `CPU_CORTEX_M` is that arch/arm/core directly makes use of CMSIS-Core in places like the following:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/f79fbac101f8febb459acee577e9f6b76e00acc0/arch/arm/core/prep_c.c#L91-L97
> 
> Since `CPU_CORTEX_M` signifies that ARM Cortex-M arch is enabled and this arch implementation depends on CMSIS-Core, `HAS_CMSIS_CORE` should be selected by it.

But that's my point @stephanosio : arch/arm/ code requires **some sort of** CMSIS compilant headers/defines, but not necessarily the ARM-provided CMSIS; maybe we could have custom-defined. I thought that is the point of this Kconfig option (HAS_CMSIS): to signify that the vendor headers pull-in CMSIS headers and not own-defined.

This is just my thought, however; it seems the option has not been actually used like this.

Now: if we accept that, by default, all ARM vendor headers must pull-in the `right` ARM CMSIS, we could even depreciate the option.

stephanosio commented:
> Now: if we accept that, by default, all ARM vendor headers must pull-in the `right` ARM CMSIS, we could even depreciate the option.

I think that would be the safest approach (i.e. require all vendor HALs to use the CMSIS-Core that is included in the Zephyr repository). If the original HAL provided by a vendor includes a subset of ARM CMSIS-Core, just remove that portion in the imported HAL that and make the necessary changes to use the official CMSIS-Core instead.

The rationale is as follows:

1. If any, the official ARM CMSIS-Core should be more complete and up-to-date than the one that is provided by the vendor.
2. Making a vendor HAL use the ARM CMSIS-Core should be very easy (after all, that is the whole point of having "Cortex Microcontroller Software Interface Standard").
3. Having different versions of CMSIS that can be used, given that two CMSISes cannot coexist for the same project, will likely introduce compatibility issues in the future (e.g. arch/arm expects a certain flavour of CMSIS but HAL provides a different flavour).
4. When ARM CMSIS-DSP is added in the future, since it makes heavy use of ARM CMSIS-Core, vendor HAL's CMSIS-Core probably won't be complete enough to support ARM CMSIS-DSP.

galak commented on ext/hal/cmsis/Kconfig:
```diff
@@ -1,8 +1,10 @@
-#
 
-config HAS_CMSIS
+config HAS_CMSIS_CORE
```
So instead of:

```
config HAS_CMSIS_CORE_M
	bool
	depends on HAS_CMSIS_CORE
```

Let's have:

```
if HAS_CMSIS_CORE

config HAS_CMSIS_CORE_M
	bool

endif
```

stephanosio replied:
Done

galak reviewed (changes requested):
Let's make a slight tweak to the Kconfig
